pkg/util/volume_util_windows.go

```diff
@@ -0,0 +1,89 @@
+//go:build windows
+// +build windows
+
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"fmt"
+)
+
+var _ VolumeUtil = &volumeUtil{}
+
+type volumeUtil struct {
+	// csiProxy is the CSIProxy implementation (either v1 or v1beta)
+	csiProxy CSIProxy
+}
+
+// NewVolumeUtil returns a VolumeUtil object for performing local filesystem operations
+func NewVolumeUtil() (VolumeUtil, error) {
+	csiProxy, err := NewCSIProxy()
+	if err != nil {
+		return nil, err
+	}
+	return &volumeUtil{
+		csiProxy,
+	}, nil
+}
+
+// GetFsCapacityByte returns capacity in bytes about a mounted filesystem.
+// mountPath is the pathname of any file within the mounted filesystem. Capacity
+// returned here is total capacity.
+func (u *volumeUtil) GetFsCapacityByte(mountPath string) (int64, error) {
+	volumeID, err := u.csiProxy.GetVolumeId(mountPath)
+	if err != nil {
+		return 0, err
+	}
+	totalBytes, _, err := u.csiProxy.GetVolumeStats(volumeID)
+	if err != nil {
+		return 0, err
+	}
+	return totalBytes, nil
+}
+
+// DeleteContents deletes all the contents under the given directory
+func (u *volumeUtil) DeleteContents(mountPath string) error {
+	volumeID, err := u.csiProxy.GetVolumeId(mountPath)
+	if err != nil {
+		return err
+	}
```

Comment:
On Linux, we support subdirectories of a filesystem that are bind mounted. Will this work on Windows or will we reformat the entire disk?

Reply:
In Windows the mount powershell commands use a volume and multiple logical volumes can be created in a disk, the creation of the volumes is left to the users and could happen during the node bootstrap, [the Google Cloud docs talks about creating a single partition with the max size](https://cloud.google.com/kubernetes-engine/docs/how-to/persistent-volumes/local-ssd#formatting_local_ssds_for_windows_server_clusters) but I believe they can do something like:

```
# format a 10GB disk into 10 partitions, execute this command 10 times
PS C:\disks> Get-Disk -Number 1 | New-Partition -Size 1073741824
PS C:\disks> Get-Disk -Number 1 | Get-Partition

   DiskPath: \\?\scsi#disk&ven_msft&prod_virtual_disk#2&1f4adffe&0&000001#{53f56307-b6bf-11d0-94f2-00a0c91efb8b}

PartitionNumber  DriveLetter Offset                                     Size Type
---------------  ----------- ------                                     ---- ----
2                            16777216                                   1 GB Basic
3                            1090519040                                 1 GB Basic
4                            2164260864                                 1 GB Basic
5                            3238002688                                 1 GB Basic
6                            4311744512                                 1 GB Basic
7                            5385486336                                 1 GB Basic
8                            6459228160                                 1 GB Basic
9                            7532969984                                 1 GB Basic
10                           8606711808                                 1 GB Basic

PS C:\disks> Get-Disk -Number 1 | Get-Partition | Get-Volume | Format-Volume -Filesystem ntfs -Confirm:$false

DriveLetter FriendlyName FileSystemType DriveType HealthStatus OperationalStatus SizeRemaining    Size
----------- ------------ -------------- --------- ------------ ----------------- -------------    ----
                         NTFS           Fixed     Healthy      OK                   1007.15 MB 1024 MB
                         NTFS           Fixed     Healthy      OK                   1007.15 MB 1024 MB
                         NTFS           Fixed     Healthy      OK                   1007.15 MB 1024 MB
                         NTFS           Fixed     Healthy      OK                   1007.15 MB 1024 MB
                         NTFS           Fixed     Healthy      OK                   1007.15 MB 1024 MB
                         NTFS           Fixed     Healthy      OK                   1007.15 MB 1024 MB
                         NTFS           Fixed     Healthy      OK                   1007.15 MB 1024 MB
                         NTFS           Fixed     Healthy      OK                   1007.15 MB 1024 MB
                         NTFS           Fixed     Healthy      OK                   1007.15 MB 1024 MB

PS C:\disks> (Get-Disk -Number 1 | Get-Partition | Get-Volume).UniqueId
\\?\Volume{452e318a-5cde-421e-9831-b9853c521012}\
\\?\Volume{e56d239c-f207-41d8-b942-111bbde76516}\
\\?\Volume{c148a6bc-22cb-49af-9119-3549ddf302c5}\
\\?\Volume{a0c91227-1d85-4836-83a1-0dac975ef81d}\
\\?\Volume{e978bf1a-1d2f-4a0e-8b41-600a774364f3}\
\\?\Volume{5db13a97-94d4-45d3-b20d-59b20dac7c97}\
\\?\Volume{a677f928-4aa6-483b-8937-d152534f9c65}\
\\?\Volume{548eb1fa-02bb-4f15-aec9-a9e95902f331}\
\\?\Volume{d01b7d64-ce64-4a54-9add-8e9b6ec238f9}\
```

The format API in CSI Proxy receives the volume's `UniqueId` shown above and will only reformat the volume and not the entire disk so I think we can do something similar to what you say.

Reply:
Ok it sounds like in Windows, we only support partitions. Can we update the docs and add a Windows section?

Reply:
sure, I created https://github.com/kubernetes-sigs/sig-storage-local-static-provisioner/issues/278